### DISCUSSION TITLE
Asset that runs a shell script inside a Docker container

### DISCUSSION THREAD

**Discussion by nx10** [dagster-io/dagster#17100] [Q&A] [answered]
Hello everyone,

We are currently in the process of assessing the feasibility of migrating our neuroimaging pipeline, which is housed at [our GitHub repository](https://github.com/FCP-INDI/C-PAC), to Dagster.

The majority of our neuroimaging tools are designed as somewhat non-portable shell applications that interact with disk I/O. To address this, we are exploring the creation of a Dagster asset that essentially orchestrates `docker run --rm` commands, utilizing distinct Docker images for each tool. These images would dynamically specify a data directory to be mounted and execute dynamically generated commands.

I'm encountering some challenges in grasping the online documentation for this particular use-case, and I have a few questions:

- Based on my interpretation of the dagster-docker documentation, does it presume that the Docker container already includes an installation of Dagster?
- Is there a means to define an asset that initiates and runs a shell script within a container (that does not have Dagster installed), following the description I provided above?
- Should we be considering a different tool for this specific use-case, and is Dagster perhaps not the most suitable choice?

Any insights and guidance on these matters would be greatly appreciated. Thank you!

**Comment by alangenfeld** [comment, marked as answer]
> Based on my interpretation of the dagster-docker documentation, does it presume that the Docker container already includes an installation of Dagster?

The `DockerExecutor` specifically has this expectation, which you don't need to use. 

> Is there a means to define an asset that initiates and runs a shell script within a container (that does not have Dagster installed), following the description I provided above?

you can use 
* `execute_docker_container` https://docs.dagster.io/_apidocs/libraries/dagster-docker#dagster_docker.execute_docker_container
* `DockerPipesClient` https://docs.dagster.io/_apidocs/libraries/dagster-docker#pipes context in https://github.com/dagster-io/dagster/discussions/16319
* or just call the docker python client yourself

> Should we be considering a different tool for this specific use-case, and is Dagster perhaps not the most suitable choice?

From the information here I see no reason Dagster would not be a viable option.

**Reply by nx10** [reply]
Awesome, thank you!
